### PR TITLE
add priority and collectionmode to diffractionplan

### PR DESCRIPTION
**Pull request by stufisher**
Merge of the following tickets:
https://github.com/ispyb/ispyb-database-modeling/issues/59
https://github.com/ispyb/ispyb-database-modeling/issues/60

### CODE REVIEW

**Review by KarlLevik — APPROVED** [review]
Looks good to me. (TBD with Neil, though.)

**Comment by stufisher** [comment]
@jlmuir?

**Comment by jlmuir** [comment]
At IMCA-CAT, 1 is the highest priority (e.g., in the set {1, 2, 3}, 1 is the highest priority, and 3 is the lowest).  To me this is the most natural understanding of a priority and is consistent with uses in our cultures such as a top-ten list or an expression like, "Right now, gaining home-field advantage is our number one priority."  Inverting that to make it easier to extend would only provide a minimal benefit, IMO, since it would only let you add something with a priority higher than all the others.  It wouldn't solve the problem of inserting a priority somewhere between the highest and lowest, or lower than the lowest, without needing to renumber.  There is of course always the option for a user to leave a gap between their priorities so that they could insert priorities later if they wanted to (e.g., {10, 20, 30, 40}) just like programming in BASIC. :-)  Also, the priorities are really just per shipment, so it's not like they have to last and be extensible over a long period of time.  So, I still favor 1 being the highest priority.  However, if you really wanted to invert the priority, IMCA-CAT could still work with that.

**Comment by stufisher** [comment]
@drnasmith ?

**Comment by drnasmith** [comment]
Fair enough. I care more that the intended use is documented than which way round it is. 1 as highest is ok with me - especially as its relative to a shipment and not an absolute measure,

**Comment by rjgildea** [comment]
I have no particularly strong opinion as to whether 1 should be highest or lowest priority, however I do note that elsewhere we already have the precendence that larger number == more important in the `AutoProcProgramAttachment.importanceRank` column:

https://github.com/DiamondLightSource/ispyb-database/blob/73beda12d868021a118d0627cb1478f9f5c74fc9/docs/list_of_tables_and_columns.rst#L100

It would probably make sense to be self-consistent here.

As a counter-example, Python logging levels also use the concept that bigger numbers are more important:

https://docs.python.org/3/library/logging.html#logging-levels

**Comment by jlmuir** [comment]
The `importanceRank` thing is unfortunate, IMO, but I agree that self-consistency is more important.  In a perfect world, I would change `importanceRank` to make 1 the highest priority.  But I know that's unrealistic with all the users of that column already, so sadly, for this new `priority` column, I agree that 1 should be the _lowest_ priority to be consistent with the existing `importanceRank` column.

**Comment by rjgildea** [comment]
I think `importanceRank` was originally proposed to be `1` as "most important", and then reversed for reasons discussed above. I'm not sure this has made it in to the official ISPyB yet, only into the Diamond version of the database. I don't think we actually use it yet, so we could still revert the definition to `1` being the most important if we think that is preferable.

https://github.com/ispyb/ispyb-database-modeling/issues/43

https://jira.diamond.ac.uk/browse/SCI-9300

**Comment by jlmuir** [comment]
> I don't think we actually use it yet, so we could still revert the definition

In that case, I would wish for the `importanceRank` column to be changed to use 1 as the highest priority and of course this new `priority` column to use 1 as the highest priority.

I can see something like a logging level numeric value having a different meaning, but for anything called "priority" or "rank", I feel like 1 naturally means highest.  For example, see [Chess Ratings](https://www.chess.com/ratings): the rank of the highest-ranked player is 1, not some number that only has meaning relative to the others.  (Not to be confused with rating which is like a score and hence has 0 as the lowest rating.)

File: schema/updates/20200822_diffractionplan_aps.sql [hunk 1]
```diff
@@ -0,0 +1,7 @@
+INSERT INTO SchemaStatus (scriptName, schemaStatus) VALUES ('20200822_diffractionplan_aps.sql', 'ONGOING');
+
+ALTER TABLE `DiffractionPlan`
+	ADD `collectionMode` ENUM('automated', 'manual') NULL COMMENT 'The requested collection mode, possible values are automated, manual',
+	ADD `priority` INT(4) NULL COMMENT 'The priority of this sample relative to others in the container';
```
Comment (jlmuir) [inline schema/updates/20200822_diffractionplan_aps.sql]: The priority is per shipment, not container.

File: docs/list_of_tables_and_columns.rst [hunk 2]
```diff
@@ -892,6 +892,8 @@
    ``monoBandwidth``,double,""
    ``centringMethod``,enum,""
    ``userPath``,varchar,"User-specified relative "root" path inside the session directory to be used for holding collected data"
+   ``collectionMode``,enum,"The requested collection mode, possible values are auto, manual"
+   ``priority``,int,"The priority of this sample relative to others in the container"
```
Comment (jlmuir) [inline docs/list_of_tables_and_columns.rst:896]: The priority is per shipment, not container.

File: schema/tables.sql [hunk 3]
```diff
@@ -2081,6 +2081,8 @@ CREATE TABLE `DiffractionPlan` (
   `monoBandwidth` double DEFAULT NULL,
   `centringMethod` enum('xray','loop','diffraction','optical') DEFAULT NULL,
   `userPath` varchar(100) DEFAULT NULL COMMENT 'User-specified relative "root" path inside the session directory to be used for holding collected data',
+  `collectionMode` enum('auto','manual') DEFAULT NULL COMMENT 'The requested collection mode, possible values are auto, manual',
+  `priority` int(4) DEFAULT NULL COMMENT 'The priority of this sample relative to others in the container',
```
Comment (jlmuir) [inline schema/tables.sql:2085]: The priority is per shipment, not container.

File: schema/updates/2020_08_22_diffractionplan_aps.sql [hunk 4]
```diff
@@ -0,0 +1,7 @@
+INSERT INTO SchemaStatus (scriptName, schemaStatus) VALUES ('2020_08_22_diffractionplan_aps.sql', 'ONGOING');
+
+ALTER TABLE `DiffractionPlan`
+	ADD `collectionMode` ENUM('auto', 'manual') NULL COMMENT 'The requested collection mode, possible values are auto, manual',
+	ADD `priority` INT(4) NULL COMMENT 'The priority of this sample relative to others in the container';
```
Comment (jlmuir) [inline schema/updates/2020_08_22_diffractionplan_aps.sql:5]: The priority is per shipment, not container.

**Review by jlmuir — CHANGES_REQUESTED** [review]
The priority is per shipment, not container, so the comments and documentation should say shipment instead of container.

**Review by jlmuir — CHANGES_REQUESTED** [review]
`schema/updates/2020_08_22_diffractionplan_aps.sql` should probably be named `2020_08_22_diffractionplan_priority_and_mode.sql` or something since the change is not really associated with the APS.

**Comment by stufisher** [comment]
Rebased on master https://github.com/DiamondLightSource/ispyb-database/pull/46

**Pull request closed without merging**